### PR TITLE
Persistent tokens and hook deletion

### PR DESCRIPTION
**Pull request by achilleas-k**
This PR introduces persistent tokens and user sessions and adds support for removing hooks from GIN repositories.

## Hook deletion
Some changes to the page templates were made to support hook deletion.
A new enum is introduced to support hooks in multiple states (e.g., a hook exists but it is disabled, or misconfigured). The plan is to be able to inform the user of a bad configuration and offer to fix it.

## Persistent tokens and session IDs
The session persistence uses token files stored as Go gobs (the same format used by gin-cli). The session storage and repository cloning works using the following strategy:
- On user login, a gin-valid token is created or an existing gin-valid token is requested from the server and stored in a directory specified in the server configuration file.
    - The username is used as the name for the file.
- A session ID is created for the user and stored in a cookie. Each session ID is mapped to a user token. This is achieved by symlinking the user's session ID to their token on disk, in the token directory, in a subdirectory called `by-sessionid`.
    - The result of this is that when a user returns with a valid cookie, any request they make that involves communicating with the GIN server (e.g., listing repositories, enabling hooks) loads their token using the session ID.
- When the user creates a hook, another symlink is created, this time linking the repository name to the user's token, in a subdirectory called `by-repo`.
    - The result of this is that when a repository hook is triggered, the cloning of the repository is achieved by loading the user's token using the repository name.

Both the session ID and the repository name are base32 encoded when linking to make the filesystem friendly (ASCII).

### CODE REVIEW

File: web/hooks.go [hunk 1]
```diff
@@ -26,35 +28,56 @@ func EnableHook(w http.ResponseWriter, r *http.Request) {
 	user := vars["user"]
 	repo := vars["repo"]
 	validator := strings.ToLower(vars["validator"])
-	cfg := config.Read()
-	cookiename := cfg.Settings.CookieName
-	sessionid, err := r.Cookie(cookiename)
+	ut, err := getSessionOrRedirect(w, r)
+	if err != nil {
+		return
```
Comment (mpsonntag) [inline web/hooks.go]: The error message is potentially not logged. `getSessionOrRedirect` logs in one error case and does not in another.

File: web/hooks.go [hunk 2]
```diff
@@ -26,35 +28,56 @@ func EnableHook(w http.ResponseWriter, r *http.Request) {
...
+	hookid, err := strconv.Atoi(hookidstr)
+	if err != nil {
+		// bad hook ID (not a number): throw a generic 404
+		fail(w, http.StatusNotFound, "not found")
 		return
 	}
+
+	ut, err := getSessionOrRedirect(w, r)
+	if err != nil {
+		return
```
Comment (mpsonntag) [inline web/hooks.go]: See `getSessionOrRedirect` error comment above.

File: web/token.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+	if err != nil {
+		return err
+	}
+	encoder := gob.NewEncoder(tokenfile)
+	return encoder.Encode(ut)
+}
+
+// loadUserToken reads a token from disk using the username as filename.
+// The location is defined by config.Dir.Tokens.
+func loadUserToken(username string) (gweb.UserToken, error) {
```
Comment (mpsonntag) [inline web/token.go]: Does it makes sense to use the name `getTokenByUser` analogous to `getTokenBySession`?

File: web/token.go [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+func getTokenByRepo(repopath string) (gweb.UserToken, error) {
+	cfg := config.Read()
+	tokendir := cfg.Dir.Tokens
+	filename := filepath.Join(tokendir, "by-repo", b32(repopath))
+	return loadToken(filename)
+}
+
+// rmTokenRepoLink deletes a repository -> token link, removing our ability to
+// clone the repository.
+func rmTokenRepoLink(repopath string) error {
```
Comment (mpsonntag) [inline web/token.go]: Are tokens and session symlinks ever cleaned up?

File: web/user.go [hunk 5]
```diff
@@ -300,7 +346,7 @@ func ShowRepo(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	session, err := getSessionOrRedirect(w, r)
+	ut, err := getSessionOrRedirect(w, r)
 	if err != nil {
 		return
```
Comment (mpsonntag) [inline web/user.go]: See `getSessionOrRedirect` error comment above.

File: web/hooks.go (hunk 1 above)
Comment (achilleas-k) [reply web/hooks.go]: `getSessionOrRedirect` logs the error only if the user has a session cookie set and the server doesn't have a session with that ID. This is the second case where the error is `Invalid session found in cookie`. The second case, `No session cookie found` isn't really an error, it just occurs when a user goes to a page that requires login and has no session cookie. I don't think it's necessary to log this, but I can add it if you want.

**Pull request merged**